src/librbd/ImageWatcher.cc

```diff
@@ -0,0 +1,643 @@
+  }
+  m_image_ctx.data_ctx.aio_flush();
+
+  unlock();
+}
+
+void ImageWatcher::notify_header_update(librados::IoCtx &io_ctx,
+				        const std::string &oid)
+{
+  // supports legacy (empty buffer) clients
+  bufferlist bl;
+  ENCODE_START(NOTIFY_VERSION, NOTIFY_VERSION, bl);
+  ::encode(NOTIFY_OP_HEADER_UPDATE, bl);
+  ENCODE_FINISH(bl);
+
+  io_ctx.notify2(oid, bl, NOTIFY_TIMEOUT, NULL);
+}
+
+std::string ImageWatcher::encode_lock_cookie() const {
+  std::ostringstream ss;
+  ss << WATCHER_LOCK_COOKIE_PREFIX << " " << m_handle;
+  return ss.str();
+}
+
+bool ImageWatcher::decode_lock_cookie(const std::string &tag,
+				      uint64_t *handle) {
+  std::string prefix;
+  std::istringstream ss(tag);
+  if (!(ss >> prefix >> *handle) || prefix != WATCHER_LOCK_COOKIE_PREFIX) {
+    return false;
+  }
+  return true;
+}
+
+void ImageWatcher::retry_aio_requests() {
+  std::vector<AioRequest> lock_request_restarts;
+  {
+    Mutex::Locker l(m_aio_request_lock);
+    assert(!m_retrying_aio_requests);
+    lock_request_restarts.swap(m_aio_requests);
+    m_retrying_aio_requests = true;
+  }
+
+  for (std::vector<AioRequest>::iterator iter = lock_request_restarts.begin();
+       iter != lock_request_restarts.end(); ++iter) {
+    ldout(m_image_ctx.cct, 10) << "retrying aio request: " << iter->second
+			       << dendl;
+    iter->first(iter->second);
+  }
+
+  Mutex::Locker l(m_aio_request_lock);
+  m_retrying_aio_requests = false;
+  m_aio_request_cond.Signal();
+}
+
+void ImageWatcher::cancel_aio_requests(int result) {
+  Mutex::Locker l(m_aio_request_lock);
+  for (std::vector<AioRequest>::iterator iter = m_aio_requests.begin();
+       iter != m_aio_requests.end(); ++iter) {
+    AioCompletion *c = iter->second;
+    c->get();
+    c->lock.Lock();
+    c->rval = result;
+    c->lock.Unlock();
+    c->finish_adding_requests(m_image_ctx.cct);
+    c->put();
+  }
+  m_aio_requests.clear();
+  m_aio_request_cond.Signal();
+}
+
+int ImageWatcher::decode_response_code(bufferlist &bl) {
+  int r;
+  bufferlist::iterator iter = bl.begin();
+  DECODE_START(NOTIFY_VERSION, iter);
+  ::decode(r, iter);
+  DECODE_FINISH(iter);
+  return r;
+}
+
+void ImageWatcher::notify_released_lock() {
+  bufferlist bl;
+  ENCODE_START(NOTIFY_VERSION, NOTIFY_VERSION, bl);
+  ::encode(NOTIFY_OP_RELEASED_LOCK, bl);
+  ENCODE_FINISH(bl);
+  m_image_ctx.md_ctx.notify2(m_image_ctx.header_oid, bl, NOTIFY_TIMEOUT, NULL);
+}
+
+void ImageWatcher::notify_request_lock() {
+  bool try_lock_complete;
+  {
+    // try to lock now that we know we are not in a rados callback
+    RWLock::WLocker l(m_image_ctx.owner_lock);
+    try_lock_complete = try_request_lock();
+  }
+  if (try_lock_complete) {
+    retry_aio_requests();
+    return;
+  }
+
+  bufferlist bl;
+  ENCODE_START(NOTIFY_VERSION, NOTIFY_VERSION, bl);
+  ::encode(NOTIFY_OP_REQUEST_LOCK, bl);
+  ENCODE_FINISH(bl);
+
+  bufferlist response;
+  int r = notify_lock_owner(bl, response);
+  if (r == -ETIMEDOUT) {
```

Comment:
do we need an || r == 0 here? It looks like if the lock owner responds, nothing may try to take the lock and send the requests again